---

# Data representations for neural networks

# Geometric Interpretation

In [75]:
import numpy as np
import tensorflow as tf
import keras

import matplotlib.pyplot as plt

# Action plan

1. What are scalars, vectors, matrices, tensors
2. What is the rank, the axis, the shape of a tensor
3. Real examples of data tensors
4. What is a (mini-)batch of data
5. Tensor operations: arithmetic, slicing, broadcasting, reshaping, element-wise vs dot/matmul ops

---

# Data representations for neural networks

### Overview

- 0 dimensional: **scalars**: plain numbers;
- 1 dimensional: **vectors**:  arrays (in Python: lists/tuples);
- 2 dimensional: **matrices**: multi-dimensional arrays → NumPy/TF;
- above: **tensors**: generalisation of matrices!

##### Note

- Math: the **dimension** or **length** of the vector  is the **number of elements** in a vector. 
- Programming: the **dimension** of a tensor is the number of **axes** (or the *length* of the *shape*)!

##### Examples

<!-- ![Figure showing scalar, 1D (vector) and 2D (matrix) tensors](images/tf/tf-shape-1.png) -->
<img alt="Figure showing scalar, 1D (vector) and 2D (matrix) tensors" style="" src="https://raw.githubusercontent.com/jchwenger/DLWP/main/lectures/02/images/tf/tf-shape-1.png">

<small>Source: [TensorFlow Introduction to Tensors](https://www.tensorflow.org/guide/tensor)</small>

<!-- ![Figure showing a 3D tensor in various transpositions](images/tf/tf-shape-2.png) -->
<img alt="Figure showing a 3D tensor in various transpositions" style="" src="https://raw.githubusercontent.com/jchwenger/DLWP/main/lectures/02/images/tf/tf-shape-2.png">

<small>Source: [TensorFlow Introduction to Tensors](https://www.tensorflow.org/guide/tensor)</small>

<!-- ![Figure showing a 4D tensor](images/tf/tf-shape-3.png) -->
<img alt="Figure showing a 4D tensor" style="" src="https://raw.githubusercontent.com/jchwenger/DLWP/main/lectures/02/images/tf/tf-shape-3.png">

<small>Source: [TensorFlow Introduction to Tensors](https://www.tensorflow.org/guide/tensor)</small>

In [3]:
x = np.array(93)
print(f"{x}, shape: {x.shape}, rank: {x.ndim}, {x.size} elements") # rank: 0
print()                                                            # the shape is an empty array!

93, shape: (), rank: 0, 1 elements



In [4]:
x = np.arange(3)
print(f"{x}, shape: {x.shape}, rank: {x.ndim}, {x.size} elements") # rank: 1
print()                                                            # shape with tuple syntax: (3,)

[0 1 2], shape: (3,), rank: 1, 3 elements



In [5]:
x = np.arange(6).reshape((2,3))
print(f"{x}, shape: {x.shape}, rank: {x.ndim}, {x.size} elements") # rank: 2
print()

[[0 1 2]
 [3 4 5]], shape: (2, 3), rank: 2, 6 elements



In [6]:
x = np.arange(12).reshape((2,2,3))
print(f"{x}, shape: {x.shape}, rank: {x.ndim}, {x.size} elements") # rank: 3

[[[ 0  1  2]
  [ 3  4  5]]

 [[ 6  7  8]
  [ 9 10 11]]], shape: (2, 2, 3), rank: 3, 12 elements


In [35]:
# Now in TensorFlow!
x = tf.constant(93)                           # .size not included as a method by default in TF
print(f"{x}, shape: {x.shape}, rank: {x.ndim}, {tf.size(x)} elements") # rank: 0
print()                                                                # the shape is an empty array!

93, shape: (), rank: 0, 1 elements



In [8]:
x = tf.range(3)
print(f"{x}, shape: {x.shape}, rank: {x.ndim}, {tf.size(x)} elements") # rank: 1
print()                                                                # shape with tuple syntax: (3,)

[0 1 2], shape: (3,), rank: 1, 3 elements



In [9]:
x = tf.reshape(tf.range(6), (2,3)) # .reshape not included as a method by default in TF
print(f"{x}, shape: {x.shape}, rank: {x.ndim}, {tf.size(x)} elements") # rank: 2
print()

[[0 1 2]
 [3 4 5]], shape: (2, 3), rank: 2, 6 elements



In [10]:
x = tf.reshape(tf.range(12), (2,2,3))
print(f"{x}, shape: {x.shape}, rank: {x.ndim}, {tf.size(x)} elements") # rank: 3

[[[ 0  1  2]
  [ 3  4  5]]

 [[ 6  7  8]
  [ 9 10 11]]], shape: (2, 2, 3), rank: 3, 12 elements


A table of comparison between mathematics and programming:

|Array |  Tensor | Maths |
|:-----|:--------|:------|
| 0D | rank 0 | scalar |
| 1D | rank 1 | vector |
| 2D | rank 2 | matrix |
| 3D | rank 3 | - |

---

## Key attributes

- **Axis** = each of the dimension of the tensor;
- **Rank** = **dimension** = number of axes (and indices needed to retrieve elements);
- **Shape** = a description of the dimensions (size of each axis);
- **Type** = the data type of the contained data – `float32`, `uint8`...

### Beware!

The NumPy and TensorFlow APIs are quite similar, with yet small differences here and there, leading to confusion.

In [11]:
print("Beware!")
print()

x = np.arange(12).reshape((2,2,3))
print(f"The default NumPy type: {x.dtype}")
print()

x = tf.reshape(tf.range(12), (2,2,3))
print(f"The default TensorFlow type: {x.dtype}")

Beware!

The default NumPy type: int64

The default TensorFlow type: <dtype: 'int32'>


### Table

|                        |        |
|-----------------------:|:-------|
| multidimensional array | tensor |
| dimension              | axis   |
| number of dimensions   | rank  
| length | number of elements in a 1D array/along a tensor axis |
| size | total number of elements (= rows * cols, for a matrix) |
| shape | lengths of each axis |

---

## Manipulating tensors in NumPy (slicing)

Data frames are pulled from tensors by an operation known as **slicing**.

This is one of the things you cannot do in plain Python (you'd need loops and they are **slow**)!

### Examples

In [12]:
# a rank 3 tensor
x = np.array([[[1, 2],
               [3, 4]],

              [[5, 6],
               [7, 8]],

              [[9, 10],
               [11, 12]]])

print(x)

[[[ 1  2]
  [ 3  4]]

 [[ 5  6]
  [ 7  8]]

 [[ 9 10]
  [11 12]]]


In [13]:
print(x[0])       # first element of the first axis

[[1 2]
 [3 4]]


In [14]:
print(x[0, 0])    # first row of the above

[1 2]


In [15]:
print(x[0, 0, 0]) # first element of the above

1


In [16]:
x = np.arange(12).reshape((3,2,2))
print(x)

[[[ 0  1]
  [ 2  3]]

 [[ 4  5]
  [ 6  7]]

 [[ 8  9]
  [10 11]]]


In [17]:
               # for each of the blocks, the first row
print(x[:, 0]) # ':' means "everything in this axis/dimension"

[[0 1]
 [4 5]
 [8 9]]


In [18]:
# for each of the blocks, the first column
# (for each block, for each column, the first element)
print(x[:, :, 0])

[[ 0  2]
 [ 4  6]
 [ 8 10]]


In [19]:
print(x[..., 0])  # same as above: '...' means: "fill in the rest"

[[ 0  2]
 [ 4  6]
 [ 8 10]]


---

## The notion of data batches

Slicing is particularly present when preparing data.

Datasets are rarely processed in one go, it's intractable.

Instead, data is broken up into **mini-batches**:

- `train_images[:128]` a mini-batch of the first 128 images;
- `train_images[128:256]` the second mini-batch;
- `train_images[128 * n, 128 * (n + 1)]` the $n+1$'th mini-batch.

Slicing MNIST image tensors about the first axis selects contiguous images

In [2]:
(train_images, train_labels), (test_images, test_labels) = tf.keras.datasets.mnist.load_data()
print("'[60'000 images, width, height]'")
print("shape", train_images.shape)  # each axis & number of elements
print("rank:", train_images.ndim)   # rank = ndim = len(shape)
print("dtype:", train_images.dtype) # data type

'[60'000 images, width, height]'
shape (60000, 28, 28)
rank: 3
dtype: uint8


In [3]:
my_slice = train_images[10:20]   # select images 10 through to image 19
print(my_slice.shape)

(10, 28, 28)


### Slicing can happen *inside* the data

Imagine we wish to extract the middle 14 by 14 sub-image of every test image. Easy with slicing!

`7:-7` from element 7 **up to, but not including**, element 21 (= 28 - 7).


0, ..., 6,|7, ..., 20|, 21, ..., 27

In [22]:
# extract the middle 14 x 14 subimages from all test images
                 # slicing width, height
my_slice = train_images[:, 7:-7, 7:-7] # : on its own means 'everything'
my_slice.shape

(60000, 14, 14)

---

## Real-world examples of data tensors

Let's look at some examples of real data tensors.

### Rank 2

Say we collected data on a population – for example, postcode, age, income.

The samples axis, the first axis, is each person surveyed and for each person we collect a vector of information, and the second axis is the data taken as a whole.

|  Data  | Rank | Shape     | Example    |
|---:|:---:|:---:|:----|
| vector | 2 | `(num_samples,num_features)` | demographics of a population - (postcode, age, income,...) for each person |

### Rank 3: time series, sequences (text/music)

Example of time series data: a data-set of weather readings.

A series of measurements – precipitation, temperature, etc. – at ten minutes intervals is collected for each day or sample.

|  Data  | Rank | Shape     | Example    |
|---:|:---:|:---:|:----|
| timeseries or sequence| 3 | `(num_samples,num_timesteps,num_features)` | daily weather features at ten minute intervals – (precipitation, temperature, humidity) for each time for each day |

### Rank 4: colour images

A colour image data-set is an example of rank 4 information.

An RGB vector resides at each pixel. The samples are individual images.

|  Data  | Rank | Shape     | Example    |
|---:|:---:|:---:|:----|
| image | 4 | `(num_samples,width,height,channels)` | 10,000 colour 256 x 256 colour images – a (256, 256, 3) colour map for each image|

##### Note

In some formats (PyTorch) you encounter  `(num_samples,channels,width,height)` instead! 💀

### Rank 5: video

About as high as you will ever meet.

The samples are videos: a video is a time series of frames and each frame is a 2D colour image.

|  Data  | Rank | Shape     | Example    |
|---:|:---:|:---:|:----|
| video | 5 | `(num_samples,num_frames,width,height,channels)` | video – a (256, 256, 3) colour map for each frame of each video|

---

# The gear of neural networks: Tensor operations

- Element-wise operations
- Broadcasting
- Tensor product
- Reshaping

---

## Element-wise operations

Element-wise means: per-element operation.

- Shapes must be **the same**;  
- **Or** compatible so that unequal axes can be **broadcast**!

Element-wise addition:

\begin{eqnarray}
\left(\begin{array}{c} 3 \\ 4 \end{array} \right)
+ \left( \begin{array}{c} 8 \\ 9 \end{array} \right)
= \left( \begin{array}{c} 11 \\ 13 \end{array} \right)
\end{eqnarray}

Hadamard product (point-wise multiplication):

\begin{eqnarray}
\left(\begin{array}{c} 3 \\ 4 \end{array} \right)
\odot \left( \begin{array}{c} 8 \\ 9 \end{array} \right)
= \left( \begin{array}{c} 24 \\ 36 \end{array} \right)
\end{eqnarray}

In [23]:
x = np.array([3, 4])
y = np.array([8, 9])
print(x + y)

[11 13]


Element-wise function application: the function is applied to each element.

\begin{eqnarray}
  f\left( \begin{array}{c} -2 \\ 3 \end{array} \right)
  = \left( \begin{array}{c} f(-2) \\ f(3) \end{array} \right)
\end{eqnarray}


For instance, the `ReLU` activation:

$$
\begin{eqnarray}
  \texttt{relu}\left( \begin{array}{c} -2 \\ 3 \end{array} \right)
   = \left( \begin{array}{c} \texttt{relu}(-2) \\ \texttt{relu}(3) \end{array} \right)
   = \left( \begin{array}{c} 0 \\ 3 \end{array} \right)
\end{eqnarray}
$$

In [33]:
# simple definition of a relu function for rank 1 tensors
def relu(x):
    assert len(x.shape) == 1
    y = x.copy()
    for i in range(x.shape[0]):
        y[i] = max(y[i], 0.)
    return y

In [34]:
x = np.array([-2, 3])
print(relu(x))
print(tf.nn.relu(x))

[0 3]
tf.Tensor([0 3], shape=(2,), dtype=int64)


**Important: *element-wise* stands in opposition to the *tensor product*, coming in a minute.**

---

## Broadcasting

**General broadcasting rules**

1. Tensor shapes are compared right to left: **right-align the axes**;
2. If needed, **add empty dimensions** (1s) to the left so that the **ranks** are equal.
3. a. Dimensions are compatible if they are **equal** or **one of them is 1**;  
   b. The axis with dimension 1 is **copied** ("stretched", "broadcast") to match the other.

<!-- ![Numpy broadcasting: 1D array with scalar -> 1D array](images/tf/np_broadcasting_1.png) -->
<img alt="Numpy broadcasting: 1D array with scalar -> 1D array" style="" src="https://raw.githubusercontent.com/jchwenger/DLWP/main/lectures/02/images/tf/np_broadcasting_1.png">

<small>Source: [NumPy Broadcasting](https://numpy.org/doc/stable/user/basics.broadcasting.html)</small>

<!-- ![Numpy broadcasting: 2D array with 1D array -> 2D array](images/tf/np_broadcasting_2.png) -->
<img alt="Numpy broadcasting: 2D array with 1D array -> 2D array" style="" src="https://raw.githubusercontent.com/jchwenger/DLWP/main/lectures/02/images/tf/np_broadcasting_2.png">

<small>Source: [NumPy Broadcasting](https://numpy.org/doc/stable/user/basics.broadcasting.html)</small>

This does not work, why? Because you first **align the dimension to the right**: `(4,3)` and `(4,) → (1,4)`!

<!-- ![Numpy broadcasting mismatch: 2D array (4x3) with 1D (4) -> incompatible (3 is not equal to 4, and neither is 1)](images/tf/np_broadcasting_3.png) -->
<img alt="Numpy broadcasting mismatch: 2D array (4x3) with 1D (4) -> incompatible (3 is not equal to 4, and neither is 1)" style="" src="https://raw.githubusercontent.com/jchwenger/DLWP/main/lectures/02/images/tf/np_broadcasting_3.png">

<small>Source: [NumPy Broadcasting](https://numpy.org/doc/stable/user/basics.broadcasting.html)</small>

Double broadcasting: `(4,1)` and `(3,) → (1,3)`

<!-- ![Numpy broadcasting: 2D array (4x1) with 1D array (3) works: result is 2D (4x3)](images/tf/np_broadcasting_4.png) -->
<img alt="Numpy broadcasting: 2D array (4x1) with 1D array (3) works: result is 2D (4x3)" style="" src="https://raw.githubusercontent.com/jchwenger/DLWP/main/lectures/02/images/tf/np_broadcasting_4.png">

<small>Source: [NumPy Broadcasting](https://numpy.org/doc/stable/user/basics.broadcasting.html)</small>

<!-- ![Sasha Rush broadcasting rules: 1) dims right aligned; 2) assign 1s to missing dimensions; 3) check compatibility, dims must be the same or one must be one; 4) broadcast by copying dims that are 1 to the number of the other dim](images/tf/srush-broadcasting.png) -->
<img alt="Sasha Rush broadcasting rules: 1) dims right aligned; 2) assign 1s to missing dimensions; 3) check compatibility, dims must be the same or one must be one; 4) broadcast by copying dims that are 1 to the number of the other dim" style="" src="https://raw.githubusercontent.com/jchwenger/DLWP/main/lectures/02/images/tf/srush-broadcasting.png">

<small>Source: [Sasha Rush, Twitter](https://twitter.com/srush_nlp/status/1516781757596680194?t=RwVp5kUWPvHG-e42wo0ryw&s=19)</small>

##### Beware!

This is an extremely useful functionality, but *sometimes* that's not what you want!

As a rule of thumb, **you cannot be too careful** about your shapes and tensor operations.

**It takes practice.**

---

## Tensor product

The dot operation is the tensor analogue of the dot product.

- between vectors, the familiar dot or scalar product:

\begin{eqnarray}
\left( \begin{array}{c} 2 & 3 \end{array} \right)
\cdot
\left( \begin{array}{c}  -1  \\ 2  \end{array} \right) =
4
\end{eqnarray}

In [26]:
x = np.array([2,3])
y = np.array([-1,2])
print(x.dot(y))
print(tf.reduce_sum(x*y))

4
tf.Tensor(4, shape=(), dtype=int64)


- between a matrix and a vector:

\begin{eqnarray}
\left( \begin{array}{cc} 2 & 3  \\ 4 & 5  \end{array} \right)
\cdot
\left( \begin{array}{c}  -1  \\ 2  \end{array} \right) =
\left( \begin{array}{c}  4  \\ 6  \end{array} \right)
\end{eqnarray}

In [27]:
x = np.array([[2,3],[4,5]])
y = np.array([-1,2])
print(x.dot(y))
print(x @ tf.reshape(y, (-1,1))) # turn y into a column vector

[4 6]
tf.Tensor(
[[4]
 [6]], shape=(2, 1), dtype=int64)


- between matrices:

\begin{eqnarray}
\left( \begin{array}{cc} 2 & 3  \\ 4 & 5  \end{array} \right)
\cdot
\left( \begin{array}{cc}  -1  & 2 \\ 2 & -3  \end{array} \right) =
\left( \begin{array}{cc}  4 & -5  \\ 6 & -7  \end{array} \right)
\end{eqnarray}

In [28]:
x = np.array([[2,3],[4,5]])
y = np.array([[-1,2], [2, -3]])
print(x.dot(y))
print(x @ y)

[[ 4 -5]
 [ 6 -7]]
[[ 4 -5]
 [ 6 -7]]


### Note: inner dimensions must match!

$$
( { \color{blue} m } \times { \color{orange} n })\cdot ( {\color{orange} n } \times { \color{green}  k } ) \rightarrow ( { \color{blue} m } \times { \color{green} k } )
$$

---

## Tensor reshaping

As we have seen, tensors elements can be redistributed in an operation known as **reshaping**.

For instance, the original `(60000, 28, 28)` MNIST images were flattened to `(60000, 28*28)`:

In [29]:
train_images_reshaped = train_images.reshape((60000, 28 * 28))
print(f"train_images shape: {train_images.shape}, after reshape: {train_images_reshaped.shape}")

train_images shape: (60000, 28, 28), after reshape: (60000, 784)


**Transposition** is another possibility.

Transposition flips rows and columns along the main diagonal.

Used all the time to make tensor dimensions agree for the tensor product!

<!-- <img alt="GIF showing the transposition of a matrix" style="height: 500px; float:right" src="images/matrix-transpose.wiki.gif"> -->
<img alt="GIF showing the transposition of a matrix" style="height: 500px; float:right" src="https://raw.githubusercontent.com/jchwenger/DLWP/main/lectures/02/images/matrix-transpose.wiki.gif">


<small style="position: absolute; bottom:0; right:0">[Transpose, Wikipedia](https://en.wikipedia.org/wiki/Transpose)</small>  

In [30]:
x = np.arange(6).reshape((2,3))
print(x)
print()
print(x.T) # same as x.transpose()
print()

x = tf.reshape(tf.range(6), (2,3))
print(x)
print()
print(tf.transpose(x)) # as before, TF has functional equivalents
                       # no x.T in TF!

[[0 1 2]
 [3 4 5]]

[[0 3]
 [1 4]
 [2 5]]

tf.Tensor(
[[0 1 2]
 [3 4 5]], shape=(2, 3), dtype=int32)

tf.Tensor(
[[0 3]
 [1 4]
 [2 5]], shape=(3, 2), dtype=int32)


# Recap

1. What are scalars, vectors, matrices, tensors  
2. What is the rank, the axis, the shape of a tensor  
3. Real examples of data tensors  
4. What is a (mini-)batch of data  
5. Tensor operations: arithmetic, slicing, broadcasting, reshaping, element-wise vs dot/matmul ops  

**Practice Makes Perfect!**

---

# Dense Layers

We have seen the fundamental operation:

$$
\bbox[5px,border:2px solid red]{
    \sigma(Wx + b) = a
}
$$

$W$: weight (matrix)  
$x$: input  
$b$: bias (vector)  
$\sigma$: activation  
$a$: output (matrix)

#### Note

[In practice](https://github.com/keras-team/keras/blob/ad27d9ccefa97ec0496ec3ffaa1df5d731acd636/keras/src/layers/core/dense.py#L212), it's implemented as ... $\sigma(xW + b)$, thank you very much...

## Graphs & Math notation

### 1. One input (two features), one neuron

$$
\newcommand\mycolv[1]{\begin{bmatrix}#1\end{bmatrix}}
   \sigma\left( \mycolv{x_{1} & x_{2}}\mycolv{w_{11}\\ w_{12}}  + b_1 \right) = a_1
$$


<!-- <img style="" src="images/nn/nn.2.svg"> -->
<img style="float:left" src="https://raw.githubusercontent.com/jchwenger/DLWP/d942b064df6d0e929cbd027bc6a6939f9e8fc763/3-getting-started/images/nn/nn.2.svg">

### 2. One input (two features), two neurons

We can do these two operations in one fell swoop using matrix multiplication.

$$
\begin{eqnarray}
\sigma \left(
\underbrace{
\underbrace{
\underbrace{
\begin{bmatrix} x_1  & x_2  \end{bmatrix}
}_{ \text{shape: } (1,2) }
\underbrace{
\begin{bmatrix} w_{11} & w_{12} \\  w_{21} & w_{22} \end{bmatrix} 
}_{ (2, 2) }
}_{ (1, 2) }
+
\underbrace{
\begin{bmatrix} b_1  & b_2 \end{bmatrix}
}_{ (1,2) }
}_{ (1,2) }
\right)
=
\underbrace{
\begin{bmatrix} a_1  & a_2 \end{bmatrix} 
}_{ (1,2) }
\end{eqnarray}
$$

<!-- <img style="" src="images/nn/nn.3.svg"> -->

<img style="float:left" src="https://raw.githubusercontent.com/jchwenger/DLWP/d942b064df6d0e929cbd027bc6a6939f9e8fc763/3-getting-started/images/nn/nn.3.svg">

### 3. One input (two features), `n` neurons

\begin{eqnarray}
\sigma \left(
\underbrace{
\underbrace{
\underbrace{
\begin{bmatrix} x_1  & x_2  \end{bmatrix}
}_{ \text{shape: }  (1, 2) }
\underbrace{
\begin{bmatrix} w_{11} & \cdots & w_{1n} \\ \\ w_{21} & \cdots  & w_{2n} \end{bmatrix}
}_{ (2, n) }
}_{ (1, n) }
+
\underbrace{
\begin{bmatrix} b_1 & \cdots & b_n \end{bmatrix} 
}_{ (1, n) }
}_{ (1, n) }
\right) = 
\underbrace{
\begin{bmatrix} a_1 & \cdots & a_n \end{bmatrix} 
}_{ (1, n) }
\end{eqnarray}


Example with $n = 12$:

<!-- <img style="" src="images/nn/nn.4.svg"> -->

<img style="float:left" src="https://raw.githubusercontent.com/jchwenger/DLWP/d942b064df6d0e929cbd027bc6a6939f9e8fc763/3-getting-started/images/nn/nn.4.svg">

### 4. Mini-batch of `m` inputs (two features), `n` neurons

\begin{eqnarray}
\sigma \left(
\underbrace{
\underbrace{
\underbrace{
\begin{bmatrix} x_{11} &  x_{12} \\ \vdots & \vdots \\ x_{m1} & x_{m2}  \end{bmatrix}
}_{ \text{shape: } (m, 2) }
\underbrace{
\begin{bmatrix} w_{11} & \cdots & w_{1n} \\ w_{21} & \cdots &  w_{2n} \end{bmatrix}
}_{ (2, n) }
}_{ (m, n) }
+
\underbrace{ \begin{bmatrix} b_1 & \cdots & b_n \end{bmatrix} }_{ (1, n) }
}_{ \text{ broadcasting: } \rightarrow (m, n) }
\right) = 
\underbrace{
\begin{bmatrix} a_{11} & \cdots & a_{1n} \\ \vdots & \ddots & \vdots \\ a_{m1} & \cdots & a_{mn} \end{bmatrix} 
}_{ (m, n)}
\end{eqnarray}


### Extra: Sum notation

The first layer of our MNIST network performed `relu(dot(w, x) + b)`.

Suppose `w` is a (512, 784) rank 2 tensor , `x` is a (784, ) rank 1 tensor and `b` is (512,).

The result of `dot(w, x)` is a (512, ) tensor.

Element `i` of `dot(w, x)` is:

\begin{align*}
(w \cdot x)_i &= \sum_{j=0}^{783} w_{ij}x_j \\
i &= 0, 1,\ldots, 511
\end{align*}

The result of `dot(w, x)` is added element-wise to `b` (they have the same shape) and `relu` is applied to each element.

The layer transformation `relu(dot(w, x) + b)` is a composition of elementary tensor operations.  

## NumPy implementation

### 1. Single input

In [59]:
# a three unit layer
n_features = 2
n_units = 4                                        # example:
x = tf.random.normal(shape=(1, n_features))        # (1,2)
w = tf.random.normal(shape=(n_features, n_units))  # (2,4) | (2,1) @ (2,4) → (2,4)
b = tf.random.normal(shape=(n_units,))             # (4,)  | (2,4) + (4,) → (2,4)

In [60]:
print(x.shape)
print(x.numpy())

(1, 2)
[[-0.3827687  1.073073 ]]


In [61]:
print(w.shape)
print(w.numpy())

(2, 4)
[[-1.2287264  -1.3801564   0.4588752  -0.72386575]
 [-0.46529594 -0.05570686 -0.73343885 -0.713819  ]]


In [62]:
print(b.shape)
print(b.numpy())

(4,)
[-0.04395255 -1.824693    1.409051   -0.15460077]


Now our affine transformation followed by `relu`.

In [63]:
y = x @ w # now our input x has been *projected* into an 4-dimensional space
print(y.shape)
print(y.numpy())

(1, 4)
[[-0.02897854  0.46850315 -0.9626765  -0.4889068 ]]


In [64]:
y += b # adding the bias elementwise
print(y.shape)
print(y.numpy())

(1, 4)
[[-0.07293109 -1.3561898   0.44637442 -0.6435076 ]]


In [65]:
a = tf.nn.relu(y) # applying ReLU elementwise
print(a.shape)
print(a.numpy())

(1, 4)
[[0.         0.         0.44637442 0.        ]]


#### 2. Mini-batch of inputs

In [66]:
# a three unit layer
n_features = 2
batch_size = 3
n_units = 4                                          # example:
x = tf.random.normal(shape=(batch_size, n_features)) # (3,2)
w = tf.random.normal(shape=(n_features, n_units))    # (2,4) | (3,2) @ (2,4) → (3,4)
b = tf.random.normal(shape=(n_units, ))              # (4,)  | (3,4) + (4,) → (3,4)

In [67]:
print(x.shape)
print(x.numpy())

(3, 2)
[[-0.7478592  0.4506009]
 [ 0.8273411 -0.3064153]
 [-1.1388956 -0.6804521]]


In [68]:
print(w.shape)
print(w.numpy())

(2, 4)
[[-0.9080192  -0.5993537   0.6183418   0.24517813]
 [-0.27917686  1.1274191   0.67797095  1.0029206 ]]


In [69]:
print(b.shape)
print(b.numpy())

(4,)
[ 2.4454253  -0.586603    1.2240415  -0.71546394]


Now our affine transformation followed by `relu`.

In [70]:
y = x @ w # now our input x has been *projected* into an 4-dimensional space
print(y.shape)
print(y.numpy())

(3, 4)
[[ 0.55327314  0.9562482  -0.15693828  0.2685582 ]
 [-0.6656975  -0.8413284   0.3038389  -0.10446427]
 [ 1.2241055  -0.08455344 -1.1655536  -0.96167177]]


In [71]:
y += b # adding the bias elementwise (broadcasting)
print(y.shape)
print(y.numpy())

(3, 4)
[[ 2.9986985   0.36964524  1.0671031  -0.44690573]
 [ 1.7797277  -1.4279313   1.5278804  -0.8199282 ]
 [ 3.6695309  -0.6711564   0.05848789 -1.6771357 ]]


In [72]:
a = tf.nn.relu(y) # applying ReLU elementwise
print(a.shape)
print(a.numpy())

(3, 4)
[[2.9986985  0.36964524 1.0671031  0.        ]
 [1.7797277  0.         1.5278804  0.        ]
 [3.6695309  0.         0.05848789 0.        ]]


## Keras implementation

```python
keras.layers.Dense(16, activation="relu")
``` 
implements: 

```python
output = relu(dot(w, input) + b)
```

- This layer transforms an input tensor into a 1D tensor of shape (16,) – a 16-dimensional vector.

- **Representational space** = **(hidden) units/neurons** = **spatial dimensions** = 16.

- Performs a **geometric transformation** resulting in a **point in 16 dimensional space**.

- Because we always process a **batch**, the dimension you'll see will be 2D: (batch_size, 16).

Think: 16 units → 16-dimensional output.

### 1. One sample

| parameter |  shape | note |
| --- |  :--- | :--- |
| input| `(1,10000)`  ||
| weights| `(10000,16)`  ||
| $xW$| `(1,16)` | `(1,10000)`$\times$`(10000,16)` → `(1,16)` |
| bias|`(16,)`  | added elementwise + broadcasting |
| output| `(1,16)`  | activation also elementwise |

### 2. Mini-batch

| parameter |  shape | note |
| --- |  :--- | :--- |
| input| `(batch_size,10000)`  ||
| weights| `(10000,16)`  ||
| $xW$| `(16,batch_size)` | `(batch_size,10000)`$\times$`(10000,16)` → `(batch_size,16)` |
| bias|`(16,)`  | added elementwise + broadcasting|
| output| `(batch_size,16)`  |  activation also elementwise |

### Sanity check

In [ ]:
x = tf.random.normal((3, 10000))
l = tf.keras.layers.Dense(16, activation="relu")
a = l(x)

In [73]:
# output
a.shape

TensorShape([3, 16])

In [74]:
# weights, biases
l.weights[0].shape, l.weights[1].shape

(TensorShape([10000, 16]), TensorShape([16]))

### Note: calculating the number of parameters 

$$
\bbox[5px,border:2px solid red]
{
\mathrm{number\ of\ params\ per\ layer} = \underbrace{\mathrm{input\ dim} \times \mathrm{number\ of\ units}}_{\mathrm{weights\ matrix}}
+
\underbrace{\mathrm{number\ of\ units}}_{\mathrm{bias\ vector}}
}
$$

In [76]:
model = keras.models.Sequential()
model.add(keras.Input((10000,)))
model.add(keras.layers.Dense(16, activation='relu'))
model.add(keras.layers.Dense(16, activation='relu'))
model.add(keras.layers.Dense(10, activation='softmax'))
model.summary()

Model: "sequential"

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━┓
┃ Layer (type)                    ┃ Output Shape           ┃       Param # ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━┩
│ dense_2 (Dense)                 │ (None, 16)             │       160,016 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_3 (Dense)                 │ (None, 16)             │           272 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_4 (Dense)                 │ (None, 10)             │           170 │
└─────────────────────────────────┴────────────────────────┴───────────────┘

 Total params: 160,458 (626.79 KB)

 Trainable params: 160,458 (626.79 KB)

 Non-trainable params: 0 (0.00 B)

In [77]:
# The weight and bias tensors
def print_layer_tensor_shape(layer):
    weight_params, bias_params = model.layers[layer].get_weights()
    print(layer, "\t", weight_params.shape, "\t", bias_params.shape)

print("layer", "\t", "(in, units)", "\t", "bias")
print_layer_tensor_shape(layer=0)
print_layer_tensor_shape(layer=1)
print_layer_tensor_shape(layer=2)

layer 	 (in, units) 	 bias
0 	 (10000, 16) 	 (16,)
1 	 (16, 16) 	 (16,)
2 	 (16, 10) 	 (10,)


In [78]:
# same as above
print(f"{(10000 * 16 + 16) + (16 * 16 + 16) + (16 * 10 + 10):,}")

160,458


---

# Geometric interpretation of tensor operations

Neural networks are of course transforming data into useful representations, but one way I like to think about that
is that they **transform space itself** to achieve their goals.

Indeed, numerical tensor elements can be interpreted not only as:

- coordinates in a (high-dimensional) real space;  
- but also as **linear transformations** in this space.

A function is linear if:
1. $f(x + y) = f(x) + f(y) \quad \text{(additivity)}$
2. $f(ax) = af(x) \quad \text{(homogeneity)}$

Otherwise it is nonlinear.

Intuitively, this is equivalent to:
- **squishing** or **stretching**;  
- **performing rotations** or a **shear**;  
- (all **straight lines** remain straight);  
- (and the **origin remains fixed**).

For example, a rotation:

<!-- ![Figure showing the effect of a rotation matrix visually](images/chollet/figure2.10.png) -->
<img alt="Figure showing the effect of a rotation matrix visually" style="" src="https://raw.githubusercontent.com/jchwenger/DLWP/main/lectures/02/images/chollet/figure2.10.png">

<small>[DLWP](https://deeplearningwithpython.io/chapters/chapter02_mathematical-building-blocks/#geometric-interpretation-of-tensor-operations), Chapter 2, Figure 2.10</small>

Or this:

<!-- ![Figure showing the effect of a matrix flipping and shrinking the y dimension (vertical mirroring)](images/chollet/figure2.11.png) -->
<img alt="Figure showing the effect of a matrix flipping and shrinking the y dimension (vertical mirroring)" style="" src="https://raw.githubusercontent.com/jchwenger/DLWP/main/lectures/02/images/chollet/figure2.11.png">

<small>[DLWP](https://deeplearningwithpython.io/chapters/chapter02_mathematical-building-blocks/#geometric-interpretation-of-tensor-operations), Chapter 2, Figure 2.11</small>

<!-- \begin{align*}
w\cdot x = \begin{pmatrix} \cos \theta & -\sin \theta \\ \sin \theta & \cos \theta \end{pmatrix}
\begin{pmatrix} x_1 \\ x_2 \end{pmatrix}
=
\begin{pmatrix} x_1 \cos \theta - x_2 \sin \theta \\ x_1 \sin \theta + y_2 \cos \theta \end{pmatrix}
\end{align*} -->

$w \cdot x$ is a linear transformation.

If you add a bias, $wx + b$, this is no longer linear, but **affine**: the origin does not stay where it is. (Same as linear transformations **plus** translations.)

##### Note

In the deep learning world, *often* you will only hear **linear** to describe these transformations, even if **affine** is the rigorous mathematical concept.

<!-- ![Figuring showing vector addition as translation](images/chollet/figure2.9.png) -->
<img alt="Figuring showing vector addition as translation" style="" src="https://raw.githubusercontent.com/jchwenger/DLWP/main/lectures/02/images/chollet/figure2.9.png">

<small>[DLWP](https://deeplearningwithpython.io/chapters/chapter02_mathematical-building-blocks/#geometric-interpretation-of-tensor-operations), Chapter 2, Figure 2.9</small>

<!-- ![Figure showing the combined effect of an affine transformation Wx + b](images/chollet/figure2.12.png) -->
<img alt="Figure showing the combined effect of an affine transformation Wx + b" style="" src="https://raw.githubusercontent.com/jchwenger/DLWP/main/lectures/02/images/chollet/figure2.12.png">

<small>[DLWP](https://deeplearningwithpython.io/chapters/chapter02_mathematical-building-blocks/#geometric-interpretation-of-tensor-operations), Chapter 2, Figure 2.12</small>

Each layer in a neural network performs a non-linear data transformation:

$$f(w\cdot x + b)$$

where $f$ is the **activation function** and $w \cdot x + b$ is an affine transformation.

<!-- ![Figure showing the effect of the the same affine transform, followed by a nonlinearity (the rectangle is first deformed 'linearly', then part of it is 'cut', since projected onto 0](images/chollet/figure2.13.png) -->
<img alt="Figure showing the effect of the the same affine transform, followed by a nonlinearity (the rectangle is first deformed 'linearly', then part of it is 'cut', since projected onto 0" style="" src="https://raw.githubusercontent.com/jchwenger/DLWP/main/lectures/02/images/chollet/figure2.13.png">

<small>[DLWP](https://deeplearningwithpython.io/chapters/chapter02_mathematical-building-blocks/#geometric-interpretation-of-tensor-operations), Chapter 2, Figure 2.13</small>

The network as a whole, in a series of linear, affine and non-linear transformations  implements a *complex transformation on a high dimensional geometric object* (the data).

The overall function of the network is **composed of a series of simpler transformations**.

The network unfolds complex data into something simple and meaningful (a representation where you can easily separate your classes...). Chollet's image: unfolding a paper sheet, crumpled into a 1000 dimensional ball.

<!-- ![Figure showing a ball of paper first crumpled, then uncrumpled, as an image for the effect of a neural network on data](images/chollet/figure2.14.png) -->
<img alt="Figure showing a ball of paper first crumpled, then uncrumpled, as an image for the effect of a neural network on data" style="" src="https://raw.githubusercontent.com/jchwenger/DLWP/main/lectures/02/images/chollet/figure2.14.png">

<small>[DLWP](https://deeplearningwithpython.io/chapters/chapter02_mathematical-building-blocks/#the-gears-of-neural-networks-tensor-operations), Chapter 2, Figure 2.14</small>